go-controller/pkg/ovn/dns-name-resolver/egressfirewall_external_dns.go

```diff
@@ -0,0 +1,386 @@
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// DNSNameResolver object was deleted.
+
+			// Get the corresponding DNS name, if avialable.
+			dnsName, found := extEgDNS.resolverToDNSName[name]
+			// If the DNS name was not found then return.
+			if !found {
+				return nil
+			}
+
+			// Delete the address set associated with the DNS name.
+			var err error
+			waitErr := wait.PollUntilContextTimeout(context.Background(), 100*time.Millisecond, 1*time.Second, true, func(ctx context.Context) (bool, error) {
+				deleteResp := extEgDNS.Delete(DeleteRequest{DNSName: dnsName})
+				err = deleteResp.Err
+				if err != nil {
+					return false, nil
+				}
+				return true, nil
+			})
+			if waitErr != nil {
+				errs := []error{waitErr}
+				if err != nil {
+					errs = append(errs, err)
+				}
+				return fmt.Errorf("error with reconcileDNSNameResolver - %v", errors.NewAggregate(errs))
+			}
+
+			// Delete the corresponding entries from the objDNSName
+			// and the dnsNameObj maps.
+			delete(extEgDNS.resolverToDNSName, name)
+			delete(extEgDNS.dnsNameToResolver, dnsName)
+
+			return nil
+		}
+		return fmt.Errorf("reconcileDNSNameResolver failed to fetch dns name resolver %s in namespace %s", name, namespace)
+	}
+
+	// DNSNameResolver object was added/updated.
+
+	// Check if the object name corresponding to the DNS name exists
+	// or not. If it exists, whether the object name matches with the
+	// exisiting extry.
+	dnsName := string(obj.Spec.Name)
+	objName, exists := extEgDNS.dnsNameToResolver[dnsName]
+	if exists && objName != name {
+		return nil
+	}
+
+	// Add the detials of the object name and the DNS name to the
+	// objDNSName and the dnsNameObj maps.
+	extEgDNS.resolverToDNSName[name] = dnsName
+	extEgDNS.dnsNameToResolver[dnsName] = name
+
+	// Get the addresses corresponding to the DNS name and add them to
+	// the address set corresponding to the DNS name.
+	addresses := []string{}
+	for _, resolvedName := range obj.Status.ResolvedNames {
+		for _, resolvedAddress := range resolvedName.ResolvedAddresses {
+			addresses = append(addresses, resolvedAddress.IP)
+		}
+	}
+	addResp := extEgDNS.Add(AddRequest{DNSName: dnsName, Addresses: addresses})
+	return addResp.Err
+
+}
+
+// Add is called whenever a DNS name is needed to be added or updated.
+func (extEgDNS *ExternalEgressDNS) Add(request AddRequest) AddResponse {
+	extEgDNS.dnsLock.Lock()
+	defer extEgDNS.dnsLock.Unlock()
+
+	// Get the details of the resolved name corresponding to the DNS name.
+	resolvedName, exists := extEgDNS.dnsNames[request.DNSName]
+
+	// If the details of the resolved name is not available, then create
+	// the address set for the DNS name.
+	if !exists {
+		if extEgDNS.addressSetFactory == nil {
+			return AddResponse{
+				Err: fmt.Errorf("error adding DNS name %s: addressSetFactory is nil", request.DNSName),
+			}
+		}
+		asIndex := GetEgressFirewallDNSAddrSetDbIDs(request.DNSName, extEgDNS.controllerName)
```

Comment:
note to self: we may need to rename addrset owner to DNSName instead of egress firewall at some point